### DISCUSSION TITLE
Constant Sampling Rate of Internal ADC

### DISCUSSION THREAD

**Discussion by JohannesMaierhofer** [micropython/micropython#13442] [ESP32]
Hello, I use the following code to sample two ADC pins from the ESP32 at 10_000Hz. The sample rate is no problem but due to the structure of the code, the sample rate is not exact. Any ideas how to sample in more sound way?

I would prefer the I2S ADC version, but unfortunately it is not implemented yet. 

```
        CurrentValues=[0] * self.n_samples
        VoltageValues=[0] * self.n_samples
        
        CurrentValues[0]=adc_current.read_uv()
        VoltageValues[0]=adc_voltage.read_uv()
        
        self.r.write_pulses((1,self.params["PulseTime"]*1000,1), 0) # Send 0 for 1us, 1 for xxxus, 0 for 1us

        last = time.ticks_us()
        i=1
        while i<self.n_samples:
            if time.ticks_diff(time.ticks_us(), last) > 1/self.f_sample*1000*1000:
                CurrentValues[i]=adc_current.read_uv()
                VoltageValues[i]=adc_voltage.read_uv()
                last = time.ticks_us()
                i+=1
```

**Comment by jimmo** [comment]
@JohannesMaierhofer On other ports (e.g. stm32) we have a way to sample the ADC at a precise interval -- https://docs.micropython.org/en/latest/library/pyb.ADC.html#pyb.ADC.read_timed

Unfortunately this is not available on ESP32.

One thing you could do to improve the jitter in the code is to optimise the function:
* Don't use floating point (these result in allocations for every calculation). i.e. instead figure out the period once as an integer.
* Cache the adc_current.read_uv, adc_voltage.read_uv, time.ticks_us, and time.ticks_diff functions in local variables to avoid global and attribute lookups.

**Comment by JohannesMaierhofer** [comment]
Thanks for the tips. It helped a lot. Additionally, I moved the line `last = time.ticks_us()` directly under the if clause. The period is corrected by 5us and I'm happy with the results.
